File: sdk/python/vellaveto/claude_agent.py
"""
Anthropic Claude Agent SDK integration for Vellaveto.

Provides tool permission enforcement hooks for Claude Agent SDK applications.
Evaluates tool calls against Vellaveto policies before execution, bridging
the Agent SDK's sandbox-runtime allowlists with Vellaveto's path_rules and
network_rules.

Example:
    from claude_agent_sdk import Agent, tool
    from vellaveto import VellavetoClient
    from vellaveto.claude_agent import VellavetoToolPermission

    client = VellavetoClient(url="http://localhost:3000")
    permission = VellavetoToolPermission(client)

    @tool
    def read_file(path: str) -> str:
        return open(path).read()

    agent = Agent(
        tools=[read_file],
        tool_permission_callback=permission.check,
    )
"""

import logging
import threading
from typing import Any, Callable, Dict, List, Optional
from urllib.parse import urlparse

from vellaveto.client import ApprovalRequired, PolicyDenied, VellavetoClient
from vellaveto.types import EvaluationContext, Verdict

logger = logging.getLogger(__name__)

_PATH_KEYS = frozenset({
    "path", "file", "filepath", "file_path", "filename", "directory",
    "dir", "folder", "src", "dst", "source", "destination", "output",
    "input", "location", "target",
})
_DOMAIN_KEYS = frozenset({
    "url", "uri", "endpoint", "host", "domain", "api_url", "base_url",
    "webhook_url", "server", "address",
})

MAX_CALL_CHAIN = 20
MAX_FIELD_LENGTH = 256


class VellavetoToolPermission:
    """
    Tool permission callback for Claude Agent SDK.

    Integrates with the Claude Agent SDK's ``tool_permission_callback``
    parameter. Each tool invocation is evaluated against Vellaveto policies.
    If the policy denies the action, the SDK receives a denial response
    preventing the tool from executing.

    The permission callback bridges Claude Agent SDK's sandbox-runtime
    domain/path allowlists with Vellaveto's ``path_rules`` and
    ``network_rules`` for unified policy enforcement.

    Args:
        client: VellavetoClient instance for policy evaluation.
        session_id: Optional session identifier for audit correlation.
        agent_id: Optional agent identifier.
        tenant_id: Optional tenant identifier.
        deny_on_error: If True (default), deny on evaluation errors.
        metadata: Additional metadata for evaluation context.
    """

    def __init__(
        self,
        client: VellavetoClient,
        *,
        session_id: Optional[str] = None,
        agent_id: Optional[str] = None,
        tenant_id: Optional[str] = None,
        deny_on_error: bool = True,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> None:
        self._client = client
        self._session_id = session_id
        self._agent_id = agent_id
        self._tenant_id = tenant_id
        self._deny_on_error = deny_on_error
        self._metadata = metadata or {}
        self._call_chain: List[str] = []
        self._lock = threading.Lock()

    def _append_chain(self, entry: str) -> None:
        with self._lock:
            self._call_chain.append(entry[:MAX_FIELD_LENGTH])
            if len(self._call_chain) > MAX_CALL_CHAIN:
                self._call_chain = self._call_chain[-MAX_CALL_CHAIN:]

    def _get_chain(self) -> List[str]:
        with self._lock:
            return list(self._call_chain)

    def _build_context(
        self, agent_name: Optional[str] = None
    ) -> EvaluationContext:
        meta = dict(self._metadata)
        meta["sdk"] = "claude_agent_sdk"
        if agent_name:
            meta["claude_agent_name"] = agent_name[:MAX_FIELD_LENGTH]
        return EvaluationContext(
            session_id=self._session_id,
            agent_id=self._agent_id,
            tenant_id=self._tenant_id,
            call_chain=self._get_chain(),
            metadata=meta,
        )

    def _extract_targets(
        self, params: Dict[str, Any]
    ) -> tuple:
        paths: List[str] = []
        domains: List[str] = []
        for key, value in params.items():
            if not isinstance(value, str):
                continue
            k = key.lower()
            if k in _PATH_KEYS:
                paths.append(value)
            elif k in _DOMAIN_KEYS:
                domains.append(value)
            else:
                parsed = urlparse(value)
                if parsed.scheme in ("http", "https", "ftp") and parsed.netloc:
                    domains.append(value)
                elif parsed.scheme == "file":
                    paths.append(value)
        return paths[:100], domains[:100]

    def check(
        self,
        tool_name: str,
        arguments: Dict[str, Any],
        *,
        agent_name: Optional[str] = None,
    ) -> bool:
        """
        Check tool permission against Vellaveto policies.

        This method is designed to be used as the ``tool_permission_callback``
        parameter of the Claude Agent SDK. Returns True to allow the tool
        call, False to deny it.

        Args:
            tool_name: Name of the tool being called.
            arguments: Tool call arguments dictionary.
            agent_name: Optional agent name for context.

        Returns:
            True if the tool call is allowed, False if denied.

        Raises:
            PolicyDenied: If deny_on_error is False but policy explicitly denies.
            ApprovalRequired: If the action requires human approval.
        """
        try:
            paths, domains = self._extract_targets(arguments)
            ctx = self._build_context(agent_name=agent_name)

            result = self._client.evaluate(
                tool=tool_name,
                function="*",
                parameters=arguments,
                target_paths=paths,
                target_domains=domains,
                context=ctx,
            )

            self._append_chain(tool_name)

            if result.verdict == "allow":
                return True
            elif result.verdict == "require_approval":
                logger.info(
                    "Claude agent tool requires approval: tool=%s",
                    tool_name,
                )
                raise ApprovalRequired(result.reason, result.approval_id)
            else:
                logger.warning(
                    "Claude agent tool denied: tool=%s reason=%s",
                    tool_name,
                    result.reason,
                )
                return False
        except (PolicyDenied, ApprovalRequired):
            raise
        except Exception:
            logger.exception(
                "Error evaluating tool permission: tool=%s", tool_name
            )
            if self._deny_on_error:
                return False
            return True

    def wrap_tool(
        self,
        func: Callable[..., Any],
        *,
        tool_name: Optional[str] = None,
        agent_name: Optional[str] = None,
    ) -> Callable[..., Any]:
        """
        Wrap a tool function with Vellaveto policy enforcement.

        Use this when the Claude Agent SDK does not support
        ``tool_permission_callback`` or for pre-SDK-integration use cases.

        Args:
            func: The tool function to wrap.
            tool_name: Override tool name (defaults to function name).
            agent_name: Optional agent name for context.

        Returns:
            Wrapped function with policy enforcement.
        """
        guard = self
        _tool_name = tool_name or getattr(
            func, "__qualname__", func.__name__
        )
        _agent_name = agent_name

        def wrapper(*args: Any, **kwargs: Any) -> Any:
            if not guard.check(_tool_name, kwargs, agent_name=_agent_name):
                raise PolicyDenied(
                    f"Tool '{_tool_name}' denied by Vellaveto policy"
                )
            return func(*args, **kwargs)

        wrapper.__name__ = func.__name__
        wrapper.__doc__ = func.__doc__
        wrapper.__qualname__ = getattr(func, "__qualname__", func.__name__)
        for attr in ("__annotations__", "name", "description", "parameters"):
            if hasattr(func, attr):
                setattr(wrapper, attr, getattr(func, attr))
        return wrapper

    def wrap_tools(
        self,
        tools: List[Callable[..., Any]],
        *,
        agent_name: Optional[str] = None,
    ) -> List[Callable[..., Any]]:
        """
        Wrap multiple tools with policy enforcement.

        Args:
            tools: List of tool functions.
            agent_name: Optional agent name for context.

        Returns:
            New list with all tools wrapped.
        """
        return [self.wrap_tool(t, agent_name=agent_name) for t in tools]

    def create_allowed_tools_filter(
        self,
        available_tools: List[str],
        *,
        agent_name: Optional[str] = None,
    ) -> List[str]:
        """
        Filter a list of tool names to only those allowed by Vellaveto policies.

        Useful for populating the Claude Agent SDK's ``allowedTools``
        configuration based on the current policy state.

        Args:
            available_tools: List of tool names to filter.
            agent_name: Optional agent name for context.

        Returns:
            Filtered list of allowed tool names.
        """
        allowed = []
        empty_args: Dict[str, Any] = {}
        for tool_name in available_tools:
            try:
                if self.check(tool_name, empty_args, agent_name=agent_name):
                    allowed.append(tool_name)
            except (PolicyDenied, ApprovalRequired):
                continue
            except Exception:
                logger.debug("Skipping tool %s in filter due to error", tool_name)
                continue
        return allowed

File: sdk/typescript/src/claude-agent.ts
/**
 * Anthropic Claude Agent SDK integration for Vellaveto.
 *
 * Provides tool permission enforcement hooks for Claude Agent SDK
 * applications in TypeScript. Evaluates tool calls against Vellaveto
 * policies before execution.
 *
 * @example
 * ```ts
 * import { VellavetoClient } from "vellaveto";
 * import { VellavetoToolPermission } from "vellaveto/claude-agent";
 *
 * const client = new VellavetoClient({ baseUrl: "http://localhost:3000" });
 * const permission = new VellavetoToolPermission(client);
 *
 * // Use as tool permission callback
 * const agent = new Agent({
 *   tools: [readFile, webSearch],
 *   toolPermissionCallback: (name, args) => permission.check(name, args),
 * });
 * ```
 */

import { VellavetoClient, PolicyDenied, ApprovalRequired } from "./client";
import { EvaluationContext, Verdict } from "./types";

const PATH_KEYS = new Set([
  "path", "file", "filepath", "file_path", "filename", "directory",
  "dir", "folder", "src", "dst", "source", "destination", "output",
  "input", "location", "target",
]);

const DOMAIN_KEYS = new Set([
  "url", "uri", "endpoint", "host", "domain", "api_url", "base_url",
  "webhook_url", "server", "address",
]);

const MAX_CALL_CHAIN = 20;
const MAX_FIELD_LENGTH = 256;
const MAX_TARGETS = 100;

/** Options for VellavetoToolPermission. */
export interface ToolPermissionOptions {
  /** Session identifier for audit correlation. */
  sessionId?: string;
  /** Agent identifier. */
  agentId?: string;
  /** Tenant identifier. */
  tenantId?: string;
  /** Deny on evaluation errors (default: true, fail-closed). */
  denyOnError?: boolean;
  /** Additional metadata for evaluation context. */
  metadata?: Record<string, unknown>;
}

/**
 * Tool permission callback for Claude Agent SDK.
 *
 * Integrates with the Claude Agent SDK's `toolPermissionCallback` parameter.
 * Each tool invocation is evaluated against Vellaveto policies.
 */
export class VellavetoToolPermission {
  private readonly client: VellavetoClient;
  private readonly sessionId?: string;
  private readonly agentId?: string;
  private readonly tenantId?: string;
  private readonly denyOnError: boolean;
  private readonly metadata: Record<string, unknown>;
  private callChain: string[] = [];

  constructor(client: VellavetoClient, options: ToolPermissionOptions = {}) {
    this.client = client;
    this.sessionId = options.sessionId;
    this.agentId = options.agentId;
    this.tenantId = options.tenantId;
    this.denyOnError = options.denyOnError ?? true;
    this.metadata = options.metadata ?? {};
  }

  private appendChain(entry: string): void {
    this.callChain.push(entry.slice(0, MAX_FIELD_LENGTH));
    if (this.callChain.length > MAX_CALL_CHAIN) {
      this.callChain = this.callChain.slice(-MAX_CALL_CHAIN);
    }
  }

  private buildContext(agentName?: string): EvaluationContext {
    const meta: Record<string, unknown> = { ...this.metadata, sdk: "claude_agent_sdk" };
    if (agentName) {
      meta.claude_agent_name = agentName.slice(0, MAX_FIELD_LENGTH);
    }
    return {
      session_id: this.sessionId,
      agent_id: this.agentId,
      tenant_id: this.tenantId,
      call_chain: [...this.callChain],
      metadata: meta,
    };
  }

  private extractTargets(params: Record<string, unknown>): {
    paths: string[];
    domains: string[];
  } {
    const paths: string[] = [];
    const domains: string[] = [];

    for (const [key, value] of Object.entries(params)) {
      if (typeof value !== "string") continue;
      const k = key.toLowerCase();
      if (PATH_KEYS.has(k)) {
        paths.push(value);
      } else if (DOMAIN_KEYS.has(k)) {
        domains.push(value);
      } else if (
        value.startsWith("http://") ||
        value.startsWith("https://") ||
        value.startsWith("ftp://")
      ) {
        domains.push(value);
      } else if (value.startsWith("file://")) {
        paths.push(value);
      }
    }

    return {
      paths: paths.slice(0, MAX_TARGETS),
      domains: domains.slice(0, MAX_TARGETS),
    };
  }

  /**
   * Check tool permission against Vellaveto policies.
   *
   * Designed to be used as the `toolPermissionCallback` parameter of the
   * Claude Agent SDK. Returns true to allow, false to deny.
   *
   * @param toolName - Name of the tool being called.
   * @param args - Tool call arguments.
   * @param agentName - Optional agent name for context.
   * @returns True if allowed, false if denied.
   * @throws ApprovalRequired if the action requires human approval.
   */
  async check(
    toolName: string,
    args: Record<string, unknown>,
    agentName?: string
  ): Promise<boolean> {
    try {
      const { paths, domains } = this.extractTargets(args);
      const ctx = this.buildContext(agentName);

      const result = await this.client.evaluate(
        {
          tool: toolName,
          function: "*",
          parameters: args,
          target_paths: paths,
          target_domains: domains,
        },
        ctx
      );

      this.appendChain(toolName);

      if (result.verdict === Verdict.Allow) {
        return true;
      } else if (result.verdict === Verdict.RequireApproval) {
        throw new ApprovalRequired(
          result.reason ?? "Approval required",
          result.approval_id ?? ""
        );
      } else {
        return false;
      }
    } catch (e) {
      if (e instanceof PolicyDenied || e instanceof ApprovalRequired) {
        throw e;
      }
      if (this.denyOnError) {
        return false;
      }
      return true;
    }
  }

  /**
   * Filter available tools to only those allowed by current policies.
   *
   * Useful for populating the Claude Agent SDK's `allowedTools`
   * configuration based on the current policy state.
   *
   * @param availableTools - List of tool names to filter.
   * @param agentName - Optional agent name for context.
   * @returns Filtered list of allowed tool names.
   */
  async filterAllowedTools(
    availableTools: string[],
    agentName?: string
  ): Promise<string[]> {
    const allowed: string[] = [];
    for (const toolName of availableTools) {
      try {
        if (await this.check(toolName, {}, agentName)) {
          allowed.push(toolName);
        }
      } catch {
        // Skip tools that throw (denied or require approval)
        continue;
      }
    }
    return allowed;
  }

  /**
   * Create a wrapped tool function with policy enforcement.
   *
   * @param fn - The tool function to wrap.
   * @param toolName - Override tool name (defaults to function name).
   * @param agentName - Optional agent name for context.
   * @returns Wrapped function with policy enforcement.
   */
  wrapTool<T extends (...args: unknown[]) => unknown>(
    fn: T,
    toolName?: string,
    agentName?: string
  ): (...args: Parameters<T>) => Promise<ReturnType<T>> {
    const guard = this;
    const name = toolName ?? fn.name;

    const wrapper = async (...args: Parameters<T>): Promise<ReturnType<T>> => {
      const kwargs =
        args.length === 1 && typeof args[0] === "object" && args[0] !== null
          ? (args[0] as Record<string, unknown>)
          : {};

      const allowed = await guard.check(name, kwargs, agentName);
      if (!allowed) {
        throw new PolicyDenied(`Tool '${name}' denied by Vellaveto policy`);
      }
      return fn(...args) as ReturnType<T>;
    };

    Object.defineProperty(wrapper, "name", { value: fn.name });
    return wrapper;
  }
}
